rs/canonical_state/src/encoding/types.rs

```diff
@@ -100,11 +100,12 @@ pub struct RequestOrResponse {
 
 #[derive(Debug, Deserialize, Serialize)]
 pub struct RequestMetadata {
-    #[serde(skip_serializing_if = "Option::is_none")]
+    // TODO(MR-642): Remove `Option` from `call_tree_depth` and `call_tree_start_time`.
+    #[serde(skip_serializing_if = "is_none_or_zero")]
     pub call_tree_depth: Option<u64>,
-    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(skip_serializing_if = "is_none_or_zero")]
     pub call_tree_start_time_u64: Option<u64>,
-    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(skip_serializing_if = "is_none_or_zero")]
     pub call_subtree_deadline_u64: Option<u64>,
```

Comment:
I'm not sure that replacing `Option::is_none` with `is_none_or_zero` is safe. We are changing the encoding of a `Request` without changing the certification version.

Both `canonical_encoding_request_with_u128_cycles()` and the changing hashes in `test_backward_compatibility()` are warning us against this.

Reply:
The hash in `canonical_encoding_request_with_u128_cycles()` still changes since we are hashing something else. I don't think it's a problem but one thing I could do would be to check if the metadata is the default on the coversion to the canonical version and then use `None` instead of Some(Default::default())`. I didn't do that because that will never be the case anyway except for the odd very old call perhaps.